Implement consistent boundary heat flux computation

gassmoeller opened the pull request:
This supersedes #2660 and implements a consistent boundary heat flux computation for our postprocessors. All tests I have run suggest the method is vastly more accurate than the old postprocessor, and also slightly more accurate than #2660 (although that is likely only important for time dependent flow) in particular also for compressible flow. I attach preliminary results for the Blankenbach and King benchmarks in the benchmark folder. This PR is not completely ready, I left a couple todos in the code, and some benchmark results are still missing, but I will not get to them in the next week, and would like some feedback on the code structure and basic method already. Maybe @ian-r-rose or someone with experience in CBF methods could have a look?

I based this implementation on Gresho, P. M., Lee, R. L., Sani, R. L., Maslanik, M. K., & Eaton, B. E.
(1987). The consistent Galerkin FEM for computing derived boundary quantities
in thermal and or fluids problems. International Journal for Numerical
Methods in Fluids, 7(4), 371-394, particularly equation (30), although adapted for our time-stepping scheme.

ian-r-rose commented on source/postprocess/heat_flux_map.cc:
```diff
@@ -40,49 +43,262 @@ namespace aspect
+                                         fe_volume_values[simulator_access.introspection().extractors.temperature].value(i,q) *
+                                         JxW;
+
+                      // TODO: Do we need stabilization terms?
```
Maybe not? Since we are close to the boundary, advection is likely to be small. And the error is not accumulating across timesteps.

ian-r-rose commented on source/postprocess/heat_flux_map.cc:
```diff
@@ -40,49 +43,262 @@ namespace aspect
+          {
+            mass_matrix[i].compress(VectorOperation::add);
+
+            // Since the mass matrix is diagonal, we can just solve for the stress vector by dividing.
```
Stale comment

ian-r-rose commented on source/postprocess/heat_flux_map.cc:
```diff
@@ -40,49 +43,262 @@ namespace aspect
+                                         JxW;
+
+                      // TODO: Do we need stabilization terms?
+                      // TODO: Consider operator splitting terms
```
What do you mean by operator splitting in this context? Since there is no timestepping for this calculation I wouldn't anticipate the need for any splitting.

SiqiZhang commented:
@gassmoeller Nice job! I did some quick tests, it works out better than #2660 in our 3D spherical models.

LudovicJnnt commented:
In the graph below a thermal convection stability test with a dodecahedral initial condition (Arrial et al 2014 benchmark). In dark blue the old (OLD) method, in light blue the average heat flux (AHF) and in yellow the CBF method. The Nusselt number for the CBF method shows smoother results which are consistent with other codes (e.g. Citcom). 
So first impression is that it looks great. I will continue to re-do all the experiments of the Arrial benchmark.

![dode_nmcomp_nut-ra-3000](https://user-images.githubusercontent.com/32137895/46348028-e6194f80-c64d-11e8-93fe-36a548263354.png)

gassmoeller commented:
I updated the tests and fixed all remaining todos, this should be ready for a review. The actual number of source code line changes is probably just 300 or so, but many test results changed. And look at the updated Blankenbach and King benchmark results :smile:.

naliboff commented on source/postprocess/heat_flux_densities.cc:
```diff
@@ -135,18 +135,15 @@ namespace aspect
+                                  "consistent boundary flux method as described in "
+                                  "Gresho, P. M., Lee, R. L., Sani, R. L., Maslanik, M. K., & Eaton, B. E. (1987). "
+                                  "The consistent Galerkin FEM for computing derived boundary quantities in thermal and or fluids "
+                                  "problems. International Journal for Numerical Methods in Fluids, 7(4), 371-394."
```
Add space.

bangerth commented:
Part of #2713.

gassmoeller commented:
I addressed the remaining comments. Lets wait for the tester to see if I excluded the no-slip boundaries correctly. Ready from my side.

gassmoeller commented:
It looks like parallel tests cause floating point exceptions. I will take a look.

gassmoeller commented:
Finally, all tests fixed and the viscosity smoothing is also repaired.
